### ISSUE TITLE
readline(FH) throws a warning, <FH> does not

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#15850]
Migrated from [rt.perl.org#130673](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=130673) (status was 'rejected')

Searchable as RT130673$

**Comment by p5pRT** [comment]
### From ryan.royal@airbus.com

A bare word file handle causes the readline\(FH\) function to emit a warning&#8203;: Name "main&#8203;::FH" used only once&#8203;: possible typo at readline\_test\.pl line 4\.

\-The equivalent diamond operator \<FH> used on the same bare word handle does not evoke the warning\.

\-Using a lexically scoped file handle with readline\( $fh \) also does not evoke the warning\.

See attached\, thanks


========================== version ===========================
Strawberry Perl

C&#8203;:\\Home\\RRoyal02\\MyPrograms\\perl\\bin> perl5\.24\.0 \-v

This is perl 5\, version 24\, subversion 0 \(v5\.24\.0\) built for MSWin32\-x64\-multi\-thread

Copyright 1987\-2016\, Larry Wall


=========================== OS ============================
Windows 7 Enterprise
Intel Xeon CPU
32 GB Ram



The information in this e\-mail is confidential\. The contents may not be disclosed or used by anyone other than the addressee\. Access to this e\-mail by anyone else is unauthorised\.
If you are not the intended recipient\, please notify Airbus immediately and delete this e\-mail\.
Airbus cannot accept any responsibility for the accuracy or completeness of this e\-mail as it has been sent over public networks\. If you have any concerns over the content of this message or its Accuracy or Integrity\, please contact Airbus immediately\.
All outgoing e\-mails from Airbus are checked using regularly updated virus scanning software but you should take whatever measures you deem to be appropriate to ensure that this message and any attachments are virus free\.

**Comment by p5pRT** [comment]
### From ryan.royal@airbus.com

﻿ScanMail detected and removed a file named "readline\_bug\.zip" that violated attachment blocking policy from the original mail entity\. You can safely save or delete this replacement attachment\.

**Comment by p5pRT** [comment]
### From @iabyn

On Mon\, Jan 30\, 2017 at 08&#8203;:44&#8203;:36AM \-0800\, ROYAL@&#8203;rt\.perl\.org wrote&#8203;:
> \# New Ticket Created by  ROYAL\, Ryan 
> \# Please include the string&#8203;:  \[perl \#130673\]
> \# in the subject line of all future correspondence about this issue\. 
> \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=130673 >
> 
> 
> A bare word file handle causes the readline\(FH\) function to emit a warning&#8203;: Name "main&#8203;::FH" used only once&#8203;: possible typo at readline\_test\.pl line 4\.
> 
> \-The equivalent diamond operator \<FH> used on the same bare word handle does not evoke the warning\.


readline takes a typeglob as an argument\, while the \<> operator takes the
\*name\* of a typeglob\, So\, as the docs for readline say\,

&nbsp;    $line = \<STDIN>;
&nbsp;    $line = readline\(\*STDIN\);    \# same thing

Note that one has a '\*'\, the other doesn't\. However\, readline will accept
a bareword string\, and at runtime use it as the name of a typeglob to look
up\.

So at \*compile\* time \(at which time 'only used once' warnings are checked
for and issued\)\, 

&nbsp;    \<FH>

is compiled as

&nbsp;    readline \*FH

and a typeglob check is done; while

&nbsp;    readline FH

is compiled as 

&nbsp;    readline 'FH'

and no typeglob is involved or checked for\,

Note that 'readline \*FH' \*will\* give a warning\.

It might be possible to change it so that in 'readline FH'\, the "FH"
string is converted to a typeglob at compile time\, but personally I'm in
favour of leaving anything filehanldle/typeglob related well alone\. It has
a ton of edge\-case weird semantics inherited from perl4 days\.

> \-Using a lexically scoped file handle with readline\( $fh \) also does not evoke the warning\.

Lexicals don't give 'only used once' warnings in any situation\, not just
readline\.

\-\- 
I before E\. Except when it isn't\.

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
@iabyn - Status changed from 'open' to 'rejected'
